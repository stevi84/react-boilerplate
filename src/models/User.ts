import { Components } from '../apis/api';

export type User = Components.Schemas.User;

export const getEmptyUser = (): User => ({
  id: 0,
  tsCreate: '',
  tsUpdate: '',
  name: '',
  dateOfBirth: '',
  size: 0,
  weight: 0,
  email: '',
  phone: '',
});
