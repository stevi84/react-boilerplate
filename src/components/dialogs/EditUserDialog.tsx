import { useNavigate, useParams } from 'react-router-dom';
import { useAppDispatch, useAppSelector } from '../../reducers/Store';
import { useEffect, useState } from 'react';
import { User, getEmptyUser } from '../../models/User';
import { Form, Formik, FormikHelpers } from 'formik';
import * as Yup from 'yup';
import { isIsoDate } from '../../globals/Validations';
import { StringEdit } from '../common/StringEdit';
import { useTranslation } from 'react-i18next';
import { DateEdit } from '../common/DateEdit';
import { createUser, readUser, updateUser } from '../../thunks/UsersThunks';
import { CrudMode } from '../dataTable/DataTableInterfaces';
import { Button, Paper, Stack, Tooltip } from '@mui/material';
import CheckIcon from '@mui/icons-material/Check';
import CloseIcon from '@mui/icons-material/Close';
import { isReadingSelector, isSubmittingSelector } from '../../reducers/ApiCallsReducer';
import { Working } from '../common/Working';
import { MainLayout } from '../common/MainLayout';
import { NumberEdit } from '../common/NumberEdit';
import { Locale } from '../../globals/Translations';

export const EditUserDialog = () => {
  const isReading: boolean = useAppSelector(isReadingSelector);
  const isSubmitting: boolean = useAppSelector(isSubmittingSelector);
  const [user, setUser] = useState<User>(getEmptyUser());
  const { id } = useParams();
  const { t, i18n } = useTranslation();
  const lang: Locale = i18n.language as Locale;
  const navigate = useNavigate();
  const dispatch = useAppDispatch();

  const mode: CrudMode | undefined = id === 'new' ? CrudMode.CREATE : /^\d+$/.test(id!) ? CrudMode.UPDATE : undefined;
  useEffect(() => {
    (async () => {
      if (mode === CrudMode.UPDATE) {
        setUser(await readUser(Number(id), lang, dispatch));
      }
    })();
  }, []);

  const save = (values: User, form: FormikHelpers<User>) => {
    switch (mode) {
      case CrudMode.CREATE:
        dispatch(createUser(values, lang));
        break;
      case CrudMode.UPDATE:
        dispatch(updateUser(values, lang));
    }
    doCancel(form)();
  };

  const doCancel = (form: FormikHelpers<User>) => () => {
    form.resetForm({ values: getEmptyUser() });
    navigate('/users');
  };

  return (
    <MainLayout allowedAccessRights={['USER_EDIT']}>
      <Paper sx={{ margin: 1 }}>
        {!isReading && !isSubmitting && (
          <Formik<User>
            initialValues={user}
            enableReinitialize={true}
            validationSchema={Yup.object({
              name: Yup.string().required(t('error_empty')),
              dateOfBirth: Yup.string().required(t('error_empty')).test('isIsoDate', t('error_date'), isIsoDate),
              size: Yup.number().required(t('error_empty')).integer(t('error_integer')).typeError(t('error_integer')),
              weight: Yup.number().required(t('error_empty')).typeError(t('error_number')),
              email: Yup.string().required(t('error_empty')).email(t('error_mail')),
              phone: Yup.string().required(t('error_empty')),
            })}
            onSubmit={save}
          >
            {(form) => (
              <Form>
                <Stack direction={'column'}>
                  <StringEdit id="name" name="name" label={t('name')} />
                  <DateEdit id="dateOfBirth" name="dateOfBirth" label={t('dateOfBirth')} />
                  <NumberEdit id="size" name="size" label={t('size')} />
                  <NumberEdit id="weight" name="weight" label={t('weight')} />
                  <StringEdit id="email" name="email" label={t('email')} />
                  <StringEdit id="phone" name="phone" label={t('phone')} />
                  <Stack direction={'row'}>
                    <Tooltip
                      title={
                        !form.dirty
                          ? t('tooltip_no_changes')
                          : !form.isValid
                          ? t('tooltip_invalid')
                          : isSubmitting
                          ? t('tooltip_submitting')
                          : t('save')
                      }
                    >
                      <span>
                        <Button
                          id="button-save"
                          variant="contained"
                          color="primary"
                          onClick={form.submitForm}
                          disabled={!form.dirty || !form.isValid || isSubmitting}
                          startIcon={<CheckIcon />}
                          autoFocus={true}
                          sx={{ margin: 1 }}
                        >
                          {t('save')}
                        </Button>
                      </span>
                    </Tooltip>
                    <Tooltip title={t('cancel')}>
                      <span>
                        <Button
                          id="button-cancel"
                          variant="outlined"
                          color="primary"
                          onClick={doCancel(form)}
                          startIcon={<CloseIcon />}
                          sx={{ margin: 1 }}
                        >
                          {t('cancel')}
                        </Button>
                      </span>
                    </Tooltip>
                  </Stack>
                </Stack>
              </Form>
            )}
          </Formik>
        )}
        {(isReading || isSubmitting) && <Working isReading={isReading} isSubmitting={isSubmitting} />}
      </Paper>
    </MainLayout>
  );
};
