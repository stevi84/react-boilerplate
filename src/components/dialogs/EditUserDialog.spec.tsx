import { describe, it, expect, vi, beforeEach } from 'vitest';
import { ShallowRenderer, createRenderer } from 'react-test-renderer/shallow';
import { fireEvent, render, screen } from '@testing-library/react';
import { user1 } from '../../../test/data/User';
import { EditUserDialog } from './EditUserDialog';
// import { createUser, updateUser } from '../../reducers/UsersReducer';
// import { getEmptyUser } from '../../models/User';

vi.mock('react-i18next', () => ({
  useTranslation: () => ({
    t: (str: string) => str,
    i18n: {
      language: 'de',
      changeLanguage: vi.fn(),
    },
  }),
}));

const { navigateMock, useParamsMock } = vi.hoisted(() => ({
  navigateMock: vi.fn(),
  useParamsMock: vi.fn(),
}));
vi.mock('react-router-dom', () => ({
  useNavigate: () => navigateMock,
  useParams: useParamsMock,
}));

vi.mock('../../hooks/UseNotifier', () => ({ useNotifier: vi.fn() }));

const { useAppSelectorMock, dispatchMock } = vi.hoisted(() => ({
  useAppSelectorMock: vi.fn(),
  dispatchMock: vi.fn(),
}));
vi.mock('../../reducers/Store', () => ({
  useAppSelector: useAppSelectorMock,
  useAppDispatch: () => dispatchMock,
}));

// ist nÃ¶tig bei Verwendung der generierten Api
const { readUserMock } = vi.hoisted(() => ({ readUserMock: vi.fn() }));
vi.mock('../../reducers/UsersReducer', async () => {
  return {
    createUser: vi.fn(),
    readUser: readUserMock,
    updateUser: vi.fn(),
  };
});

describe('EditUserDialog', () => {
  beforeEach(() => {
    readUserMock.mockResolvedValue(user1);
  });

  it('should equal saved snapshot', () => {
    useAppSelectorMock.mockReturnValue(false);
    useParamsMock.mockReturnValue({ id: 1 });
    const renderer: ShallowRenderer = createRenderer();
    renderer.render(<EditUserDialog />);
    const tree = renderer.getRenderOutput();
    expect(tree).toMatchSnapshot();
  });

  it('should load user if updating', () => {
    useAppSelectorMock.mockReturnValue(false);
    useParamsMock.mockReturnValue({ id: '1' });
    expect(readUserMock.mock.calls.length).toEqual(0);
    render(<EditUserDialog />);
    expect(readUserMock.mock.calls.length).toEqual(1);
    expect(readUserMock.mock.calls[0][0]).toEqual(1);
    expect(readUserMock.mock.calls[0][1]).toEqual('de');
    expect(readUserMock.mock.calls[0][2]).toEqual(dispatchMock);
  });

  it('should not load user if creating', () => {
    useAppSelectorMock.mockReturnValue(false);
    useParamsMock.mockReturnValue({ id: 'new' });
    render(<EditUserDialog />);
    expect(readUserMock.mock.calls.length).toEqual(0);
  });

  // it('should dispatch createUser on saving new user', () => {
  //   useAppSelectorMock.mockReturnValue(false);
  //   useParamsMock.mockReturnValue({ id: 'new' });
  //   expect(dispatchMock.mock.calls.length).toEqual(0);
  //   render(<EditUserDialog />);
  //   fireEvent.change(screen.getByLabelText('name'), { target: { value: 'name' } });
  //   fireEvent.change(screen.getByLabelText('dateOfBirth'), { target: { value: '06.12.2023' } });
  //   fireEvent.change(screen.getByLabelText('size'), { target: { value: '123' } });
  //   fireEvent.change(screen.getByLabelText('weight'), { target: { value: '50.5' } });
  //   fireEvent.change(screen.getByLabelText('email'), { target: { value: 'test@test.de' } });
  //   fireEvent.change(screen.getByLabelText('phone'), { target: { value: '0123/4567890' } });
  //   fireEvent.click(screen.getByText('save'));
  //   expect(dispatchMock.mock.calls.length).toEqual(1);
  //   expect(dispatchMock.mock.calls[0][0]).toEqual(createUser(getEmptyUser(), 'de'));
  // });

  // it('should dispatch updateUser on saving existing user', () => {
  //   useAppSelectorMock.mockReturnValue(false);
  //   useParamsMock.mockReturnValue({ id: '1' });
  //   expect(dispatchMock.mock.calls.length).toEqual(0);
  //   render(<EditUserDialog />);
  //   fireEvent.change(screen.getByLabelText('name'), { target: { value: 'change' } });
  //   fireEvent.click(screen.getByText('save'));
  //   expect(dispatchMock.mock.calls.length).toEqual(1);
  //   expect(dispatchMock.mock.calls[0][0]).toEqual(updateUser(getEmptyUser(), 'de'));
  // });

  it('should navigate to users on cancel', () => {
    useAppSelectorMock.mockReturnValue(false);
    useParamsMock.mockReturnValue({ id: '1' });
    expect(navigateMock.mock.calls.length).toEqual(0);
    render(<EditUserDialog />);
    fireEvent.click(screen.getByText('cancel'));
    expect(navigateMock.mock.calls.length).toEqual(1);
    expect(navigateMock.mock.calls[0][0]).toEqual('/users');
  });

  it('should show dialog if not reading or submitting', () => {
    useAppSelectorMock.mockReturnValue(false);
    useParamsMock.mockReturnValue({ id: 1 });
    render(<EditUserDialog />);
    expect(screen.getByLabelText('name')).toBeTruthy();
    expect(screen.queryByRole('progressbar')).toBeFalsy();
  });

  it('should show working if reading', () => {
    useAppSelectorMock.mockReturnValueOnce(true).mockReturnValue(false);
    useParamsMock.mockReturnValue({ id: 1 });
    render(<EditUserDialog />);
    expect(screen.queryByLabelText('name')).toBeFalsy();
    expect(screen.getByRole('progressbar')).toBeTruthy();
  });

  it('should show working if submitting', () => {
    useAppSelectorMock.mockReturnValueOnce(false).mockReturnValue(true);
    useParamsMock.mockReturnValue({ id: 1 });
    render(<EditUserDialog />);
    expect(screen.queryByLabelText('name')).toBeFalsy();
    expect(screen.getByRole('progressbar')).toBeTruthy();
  });
});
