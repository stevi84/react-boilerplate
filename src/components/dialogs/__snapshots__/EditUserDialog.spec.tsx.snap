// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`EditUserDialog > should equal saved snapshot 1`] = `
<MainLayout>
  <ForwardRef(Paper)
    sx={
      {
        "margin": 1,
      }
    }
  >
    <Formik
      enableReinitialize={true}
      initialValues={
        {
          "dateOfBirth": "",
          "email": "",
          "id": 0,
          "name": "",
          "phone": "",
          "size": 0,
          "tsCreate": "",
          "tsUpdate": "",
          "weight": 0,
        }
      }
      onSubmit={[Function]}
      validationSchema={
        ObjectSchema {
          "_blacklist": Set {},
          "_excludedEdges": [],
          "_mutate": undefined,
          "_nodes": [
            "phone",
            "email",
            "weight",
            "size",
            "dateOfBirth",
            "name",
          ],
          "_sortErrors": [Function],
          "_typeCheck": [Function],
          "_whitelist": Set {},
          "conditions": [],
          "deps": [],
          "exclusiveTests": {},
          "fields": {
            "dateOfBirth": StringSchema {
              "_blacklist": Set {},
              "_typeCheck": [Function],
              "_whitelist": Set {},
              "conditions": [],
              "deps": [],
              "exclusiveTests": {
                "isIsoDate": false,
                "required": false,
              },
              "internalTests": {
                "nullable": [Function],
                "optionality": [Function],
                "typeError": [Function],
              },
              "spec": {
                "abortEarly": true,
                "coerce": true,
                "disableStackTrace": false,
                "nullable": false,
                "optional": false,
                "recursive": true,
                "strict": false,
                "strip": false,
              },
              "tests": [
                [Function],
                [Function],
              ],
              "transforms": [
                [Function],
              ],
              "type": "string",
            },
            "email": StringSchema {
              "_blacklist": Set {},
              "_typeCheck": [Function],
              "_whitelist": Set {},
              "conditions": [],
              "deps": [],
              "exclusiveTests": {
                "email": false,
                "required": false,
              },
              "internalTests": {
                "nullable": [Function],
                "optionality": [Function],
                "typeError": [Function],
              },
              "spec": {
                "abortEarly": true,
                "coerce": true,
                "disableStackTrace": false,
                "nullable": false,
                "optional": false,
                "recursive": true,
                "strict": false,
                "strip": false,
              },
              "tests": [
                [Function],
                [Function],
              ],
              "transforms": [
                [Function],
              ],
              "type": "string",
            },
            "name": StringSchema {
              "_blacklist": Set {},
              "_mutate": undefined,
              "_typeCheck": [Function],
              "_whitelist": Set {},
              "conditions": [],
              "deps": [],
              "exclusiveTests": {
                "required": false,
              },
              "internalTests": {
                "nullable": [Function],
                "optionality": [Function],
                "typeError": [Function],
              },
              "spec": {
                "abortEarly": true,
                "coerce": true,
                "disableStackTrace": false,
                "nullable": false,
                "optional": false,
                "recursive": true,
                "strict": false,
                "strip": false,
              },
              "tests": [
                [Function],
              ],
              "transforms": [
                [Function],
              ],
              "type": "string",
            },
            "phone": StringSchema {
              "_blacklist": Set {},
              "_mutate": undefined,
              "_typeCheck": [Function],
              "_whitelist": Set {},
              "conditions": [],
              "deps": [],
              "exclusiveTests": {
                "required": false,
              },
              "internalTests": {
                "nullable": [Function],
                "optionality": [Function],
                "typeError": [Function],
              },
              "spec": {
                "abortEarly": true,
                "coerce": true,
                "disableStackTrace": false,
                "nullable": false,
                "optional": false,
                "recursive": true,
                "strict": false,
                "strip": false,
              },
              "tests": [
                [Function],
              ],
              "transforms": [
                [Function],
              ],
              "type": "string",
            },
            "size": NumberSchema {
              "_blacklist": Set {},
              "_typeCheck": [Function],
              "_whitelist": Set {},
              "conditions": [],
              "deps": [],
              "exclusiveTests": {
                "integer": false,
              },
              "internalTests": {
                "nullable": [Function],
                "optionality": [Function],
                "typeError": [Function],
              },
              "spec": {
                "abortEarly": true,
                "coerce": true,
                "disableStackTrace": false,
                "nullable": false,
                "optional": false,
                "recursive": true,
                "strict": false,
                "strip": false,
              },
              "tests": [
                [Function],
              ],
              "transforms": [
                [Function],
              ],
              "type": "number",
            },
            "weight": NumberSchema {
              "_blacklist": Set {},
              "_typeCheck": [Function],
              "_whitelist": Set {},
              "conditions": [],
              "deps": [],
              "exclusiveTests": {},
              "internalTests": {
                "nullable": [Function],
                "optionality": [Function],
                "typeError": [Function],
              },
              "spec": {
                "abortEarly": true,
                "coerce": true,
                "disableStackTrace": false,
                "nullable": false,
                "optional": false,
                "recursive": true,
                "strict": false,
                "strip": false,
              },
              "tests": [],
              "transforms": [
                [Function],
              ],
              "type": "number",
            },
          },
          "internalTests": {
            "nullable": [Function],
            "typeError": [Function],
          },
          "spec": {
            "abortEarly": true,
            "coerce": true,
            "disableStackTrace": false,
            "nullable": false,
            "optional": true,
            "recursive": true,
            "strict": false,
            "strip": false,
          },
          "tests": [],
          "transforms": [],
          "type": "object",
        }
      }
    >
      [Function]
    </Formik>
  </ForwardRef(Paper)>
</MainLayout>
`;
